azure-pipelines.yml

```diff
@@ -1,16 +1,135 @@
-# Starter pipeline
-# Start with a minimal pipeline that you can customize to build and deploy your code.
-# Add steps that build, run tests, deploy, and more:
-# https://aka.ms/yaml
+#
+# Copyright 2013-2019 the original author or authors from the JHipster project.
+#
+# This file is part of the JHipster project, see https://www.jhipster.tech/
+# for more information.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
 
-pool:
-  vmImage: 'Ubuntu 16.04'
+jobs:
+- job: Test
+  pool:
+    vmImage: 'Ubuntu 16.04'
+  variables:
+    JHI_PROFILE: dev
+    JHI_RUN_APP: 1
+    JHI_PROTRACTOR: 0
+    # if JHI_LIB_BRANCH value is release, use the release from Maven
+    JHI_LIB_REPO: https://github.com/jhipster/jhipster.git
+    JHI_LIB_BRANCH: release
+    # if JHI_GEN_BRANCH value is release, use the release from NPM
+    JHI_GEN_REPO: https://github.com/jhipster/generator-jhipster.git
+    JHI_GEN_BRANCH: master
+    # specific config
+    SPRING_OUTPUT_ANSI_ENABLED: NEVER
+    SPRING_JPA_SHOW_SQL: false
+    JHI_DISABLE_WEBPACK_LOGS: true
+    JHI_E2E_HEADLESS: true
+    JHI_SCRIPTS: /home/vsts/generator-jhipster/test-integration/scripts
```

Comment:
can we point this to the local node_modules folder? There should be one.

Reply:
I am not sure but in case it should be consistent with $HOME variable used in `00-init-env.sh`.
Simply using $HOME does not work in the yml file.

A list of available variables is [here](https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops) and [here](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#working-with-variables) is how to use them . I tried some such as `Agent.HomeDirectory` and `Build.Repository.LocalPath` but to no avail. I think I also tried `$(Home)` but I can not remember for sure.

Maybe @pascalgrimaud can help as I think he is the one done the Azure pipelines integration?

Reply:
Yes, I will be able to help but only tonight, as I'm currently at day work :)

Reply:
Awesome!!! Thank you very much @pascalgrimaud

Reply:
Just have a quick look here, $HOME should work.

You can look at this other repo: https://github.com/hipster-labs/jhipster-daily-builds
as I put $HOME there too: https://github.com/hipster-labs/jhipster-daily-builds/blob/master/azure-pipelines-ngx-maven.yml#L40

Reply:
@pascalgrimaud Thanks for the tip. I initially used `$HOME` but that did not work in our case as it seems to point to a different directory to the one resolved in `00-init-env.sh`. I will check and give it another try

Reply:
So I tested again with no success but here's what I found.

#### Approach 1
Declaring JHI_SCRIPTS using $HOME in the `azure-pipelines.yml`: 
```yaml 
JHI_SCRIPTS: $HOME/generator-jhipster/test-integration/scripts
```
results in the variable being resolved within `10-install-jhipster-kotlin.sh` as:
```bash
JHI_SCRIPTS=$HOME/generator-jhipster/test-integration/scripts
```
which is different from $HOME resolved in `00-init-env.sh`:
```bash
JHI_HOME=/home/vsts/generator-jhipster
```

You can see the results of this build configuration at: https://jhipster.visualstudio.com/generator-jhipster/_build/results?buildId=1813

#### Approach 2
Removed `JHI_SCRIPTS` from `azure-pipelines.yml` in order to let `00-init-env.sh` to set it to the proper directory. This time the `JHI_SCRIPTS` is resolved properly within the bash scripts **BUT** azure fails to resolve it. I have tried changing bash commands in the yml file using all the following forms but none of them worked.

The first one resolves to `/11-generate-entities.sh`
```yml
bash: $(JHI_SCRIPTS)/11-generate-entities.sh
```
This one as well resolves to` /11-generate-entities.sh`, although it should work according to the [docs](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#scripts): 
```yml
bash: $JHI_SCRIPTS/11-generate-entities.sh
```
The last one is invalid syntax and the build does not even start:
```yml
bash: "$JHI_SCRIPTS"/11-generate-entities.sh
```
You can see the results of these build configurations:
1. https://jhipster.visualstudio.com/generator-jhipster/_build/results?buildId=1814
2. https://jhipster.visualstudio.com/generator-jhipster/_build/results?buildId=1816 
3. https://jhipster.visualstudio.com/generator-jhipster/_build/results?buildId=1815&view=results

#### Summary
So it seems to me that `$HOME` variable is not properly resolved within the bash scripts and I am not sure about the yml either as it never reaches the `11-generate-entities.sh` either.

I can not see another system variable to achieve the same as `$HOME`.

Reply:
@pascalgrimaud  It is of-course very possible that I am missing something very obvious so please check as well when you get some free time. Thanks

Reply:
So I think I have found the solution. Change `00-init-env.sh` to always override the value. You are doing the same [here](https://github.com/hipster-labs/jhipster-daily-builds/blob/master/test-integration/scripts/00-init-env.sh) as well.

The fault is mine as I have actually changed `00-init-env.sh` in commit https://github.com/pvliss/jhipster-kotlin/commit/db826d5a9abc01c927878b6bf1727bfcc427c951 to match that of the parent JHipster https://github.com/jhipster/generator-jhipster/blob/master/test-integration/scripts/00-init-env.sh.

It does seem like a hack though as this worked just fine with Travis. Anyways I hope this will resolve the issue now

Reply:
It's hard to find some scripts which work on both: Azure and Travis.
Both intialize differently the variables at the start of the build, that's explain the use of this `00-init-env.sh`